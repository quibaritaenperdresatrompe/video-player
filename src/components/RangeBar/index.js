import glamorous from 'glamorous'
import PropTypes from 'prop-types'
import React, { Component } from 'react'

const RangeBarContainer = glamorous.div(({ size }) => ({
  backgroundColor: 'hsla(0, 0%, 100%, 0.8)',
  margin: `calc(${size} * 2) 0`,
  opacity: '0.8',
  position: 'relative',
  ':hover': {
    opacity: '1',
  },
}))

const StyledRangeBar = glamorous.input(({ color, size }) => {
  const thumbStyle = {
    appearance: 'none',
    backgroundColor: color,
    border: 0,
    borderRadius: '50%',
    height: `calc(${size} * 5)`,
    margin: `${size} 0`,
    width: `calc(${size} * 5)`,
  }

  return {
    appearance: 'none',
    backgroundColor: 'transparent',
    border: 0,
    cursor: 'pointer',
    height: size,
    margin: 0,
    outline: 'none',
    position: 'relative',
    width: '100%',
    '::-webkit-slider-thumb': thumbStyle,
    '::-moz-range-thumb': thumbStyle,
    '::-ms-thumb': thumbStyle,
  }
})

const StyledCurrentValueBar = glamorous.div(({ color, progress, size }) => ({
  backgroundColor: color,
  height: size,
  left: 0,
  position: 'absolute',
  top: 0,
  width: progress === 100
    ? `${progress}%`
    : `calc(${progress}% + 1px)`,
}))

class RangeBar extends Component {
  constructor() {
    super()
    this.state = {
      isSetting: false,
    }
  }

  handleChange = ({ target: { value } }) => {
    this.props.setTo(parseFloat(value, 10))
    this.setState(() => ({
      isSetting: true,
    }))
  }

  render() {
    const {
      color,
      currentValue,
      maxValue,
      size,
    } = this.props

    const progress = maxValue === 0
      ? 0
      : (currentValue * 100) / maxValue

    return (
      <RangeBarContainer size={size}>
        <StyledRangeBar
          color={color}
          max={maxValue}
          onChange={this.handleChange}
          size={size}
          step='any'
          type='range'
          value={currentValue}
        />
        <StyledCurrentValueBar
          color={color}
          progress={progress}
          size={size}
        />
      </RangeBarContainer>
    )
  }
}

RangeBar.propTypes = {
  color: PropTypes.string,
  currentValue: PropTypes.number,
  maxValue: PropTypes.number,
  setTo: PropTypes.func,
  size: PropTypes.string,
}

RangeBar.defaultProps = {
  color: 'hsla(0, 0%, 100%, 1)',
  currentValue: 0,
  maxValue: 1,
  setTo: () => undefined,
  size: '0.2em',
}

export default RangeBar
