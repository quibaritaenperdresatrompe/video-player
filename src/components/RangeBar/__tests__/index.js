import RangeBar from '../';

import itRendersAllMutations from '../../../utils/jest-it-renders-all-mutations';

const mutations = [
  {
    name: 'without props',
    props: null,
  },
  {
    name: 'without currentValue and maxValue',
    props: {
      currentValue: 9,
      maxValue: 10,
    },
  },
  {
    name: 'when it is at the end',
    props: {
      currentValue: 10,
      maxValue: 10,
    },
  },
  {
    name: 'whith all props',
    props: {
      color: '#fff',
      currentValue: 10,
      maxValue: 10,
      setTo: jest.fn(),
      size: '3em',
    },
  },
];

describe(RangeBar.name, () => {
  itRendersAllMutations(RangeBar, mutations);

  describe('handlers', function scope() {
    beforeEach(() => {
      this.RangeBar = new RangeBar();
    });

    describe('handleChange', () => {
      test('it calls props.setTo', () => {
        this.RangeBar.props = {
          setTo: jest.fn(),
        };
        const event = { target: { value: 99 } };
        this.RangeBar.handleChange(event);
        expect(this.RangeBar.props.setTo).toHaveBeenCalledWith(event.target.value);
      });
    });
  });
});
