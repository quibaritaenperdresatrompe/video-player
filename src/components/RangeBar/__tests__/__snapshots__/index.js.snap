// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeBar when it is at the end it renders correctly 1`] = `
<div
  className="css-12pnkes"
>
  <input
    className="css-aav0sr"
    max={10}
    onChange={[Function]}
    size="0.2em"
    step="any"
    type="range"
    value={10}
  />
  <div
    className="css-1jwjj7r"
  />
</div>
`;

exports[`RangeBar whith all props it renders correctly 1`] = `
<div
  className="css-pletxe"
>
  <input
    className="css-1vgo5h5"
    max={10}
    onChange={[Function]}
    size="3em"
    step="any"
    type="range"
    value={10}
  />
  <div
    className="css-4xxlbu"
  />
</div>
`;

exports[`RangeBar without currentValue and maxValue it renders correctly 1`] = `
<div
  className="css-12pnkes"
>
  <input
    className="css-aav0sr"
    max={10}
    onChange={[Function]}
    size="0.2em"
    step="any"
    type="range"
    value={9}
  />
  <div
    className="css-f50oys"
  />
</div>
`;

exports[`RangeBar without props it renders correctly 1`] = `
<div
  className="css-12pnkes"
>
  <input
    className="css-aav0sr"
    max={1}
    onChange={[Function]}
    size="0.2em"
    step="any"
    type="range"
    value={0}
  />
  <div
    className="css-1hte4yv"
  />
</div>
`;
